contracts/auctions/src/contract.rs

```diff
@@ -0,0 +1,430 @@
+            return Err(ContractError::AuctionNotActive);
+        }
+
+        if auction.status != AuctionStatus::Active {
+            log!(
+                &env,
+                "Auction: Place Bid: Trying to place a bid for inactive/cancelled auction."
+            );
+            return Err(ContractError::AuctionNotActive);
+        }
+
+        if bidder == auction.seller {
+            log!(&env, "Auction Place Bid: Seller cannot place bids.");
+            return Err(ContractError::InvalidBidder);
+        }
+
+        let token_client = token::Client::new(&env, &auction.currency);
+
+        match auction.highest_bid {
+            Some(old_highest_bid) if bid_amount > old_highest_bid => {
+                // refund the previous highest bidder
+                let old_bid_info = get_highest_bid(&env, auction_id)?;
+
+                token_client.transfer(
+                    &env.current_contract_address(),
+                    &old_bid_info.bidder,
+                    &(old_bid_info.bid as i128),
+                );
+
+                // transfer the new bid amount
+                token_client.transfer(
+                    &bidder,
+                    &env.current_contract_address(),
+                    &(bid_amount as i128),
+                );
+
+                set_highest_bid(&env, auction_id, bid_amount, bidder.clone())?;
+
+                // Update auction state
+                //TODO: in case we have separate storage isn't the two below
+                //redundant?
+                auction.highest_bid = Some(bid_amount);
+                auction.highest_bidder = bidder;
+                update_auction(&env, auction_id, auction.clone())?;
+                save_auction_by_id(&env, auction_id, &auction)?;
+                save_auction_by_seller(&env, &auction.seller, &auction)?;
+
+                Ok(())
+            }
+            Some(_) => {
+                log!(&env, "Auction: Place bid: Bid not enough");
+                Err(ContractError::BidNotEnough)
+            }
+            None => {
+                // this is in case there are no previous bids
+                // transfer the new bid amount
+                token_client.transfer(
+                    &bidder,
+                    &env.current_contract_address(),
+                    &(bid_amount as i128),
+                );
+
+                set_highest_bid(&env, auction_id, bid_amount, bidder.clone())?;
+
+                // Update auction state
+                auction.highest_bid = Some(bid_amount);
+                auction.highest_bidder = bidder;
+                update_auction(&env, auction_id, auction.clone())?;
+                save_auction_by_id(&env, auction_id, &auction)?;
+                save_auction_by_seller(&env, &auction.seller, &auction)?;
+
+                Ok(())
+            }
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn finalize_auction(env: Env, auction_id: u64) -> Result<(), ContractError> {
```

Comment:
I think this function needs some refactoring. 
There are 3 cases here:
- min price not reached
- no one bidded
- the correct case

Right now it's this weird thing where you repeat 3x in code the same sections
```
           auction.status = AuctionStatus::Ended;

            save_auction_by_id(&env, auction_id, &auction)?;
            save_auction_by_seller(&env, &auction.seller, &auction)?;
            update_auction(&env, auction_id, auction)?;

            return Ok(());
```
which is the redundancy of the highest order. Plus it makes code harder to read.

Make it into a big if-else of match statement and make the common operations happend before and after that statement.

Reply:
split the logic and placed it in `utils.rs` now it should be more readable